### ISSUE TITLE
Public constants (assembly)

### ISSUE THREAD

**Issue by frisitano** [0xPolygonMiden/miden-vm#766]
We are currently able to define constants in an assembly file and use them with the `push.*` operation.  However, the constants are scoped to the module they are defined in. It would also be useful if we could export constants such that they can be imported into other modules.  We have requirements for shared constants when developing the transaction kernel.  See: https://github.com/0xPolygonMiden/miden-base/pull/63.

**Comment by vlopes11** [comment]
Regarding the implementation, would that be an alias to `push.x`? Example: the definition `const.BLK_HASH_PTR=10` and usage `const.BLK_HASH_PTR` would compile to `push.10`

If yes, it seems we could load a constant list on the fly for the compiler and store the into the assembly context. That way, whenever we see a `const.FOO`, we query the assembly context and replace the instruction. This intersects with the idea to load MASM library bundles as it would be just a sequence of constant declarations. Something like this:

https://github.com/0xPolygonMiden/miden-vm/blob/159a04f3d4648f77d664642409e521438378bd6c/assembly/src/library.rs#L75-L91

Where the directory will define also the constants

**Comment by frisitano** [comment]
> Regarding the implementation, would that be an alias to `push.x`? Example: the definition `const.BLK_HASH_PTR=10` and usage `const.BLK_HASH_PTR` would compile to `push.10`

For a usage overview I would recommended taking a look at the [docs](https://0xpolygonmiden.github.io/miden-vm/user_docs/assembly/code_organization.html#constants) to see the usage pattern of miden constants.  But for convenience I've extracted an extract of how they look in code. It's pretty much as you say but instead you can explicitly use the constant name with the `push` operation.  In the future we would like to extend this to other operations that take immediate values.  e.g. ` mem_load.CONSTANT`

```
use.std::math::u64

const.CONSTANT_1=100
const.CONSTANT_2=200

begin
    push.CONSTANT_1.CONSTANT_2
    exec.u64::checked_add
end

``` 

> If yes, it seems we could load a constant list on the fly for the compiler and store the into the assembly context. That way, whenever we see a `const.FOO`, we query the assembly context and replace the instruction.

Yeah so this is basically how it currently works - see the PR [here](https://github.com/0xPolygonMiden/miden-vm/pull/692/files). However constants are only accessible to the module they are defined in.  We should extend this so they can be imported and obey import rules.  Similar to how procedures are made available across modules

**Comment by vlopes11** [comment]
@frisitano oh I see! The constants PR already landed. All clear, thx for the clarification.

Regarding the module issue, I think that kinda makes sense. If we have something like defined in a module `foo`

```
const.CONSTANT_1=100
```

That could be usable only if we do something like this

```
use.foo::CONSTANT_1

begin
    push.CONSTANT_1
end
```

**Comment by frisitano** [comment]
> @frisitano oh I see! The constants PR already landed. All clear, thx for the clarification.
> 
> Regarding the module issue, I think that kinda makes sense. If we have something like defined in a module `foo`
> 
> ```
> export.const.CONSTANT_1=100
> ```
> 
> That could be usable only if we do something like this
> 
> ```
> use.foo::CONSTANT_1
> 
> begin
>     push.CONSTANT_1
> end
> ```

@vlopes11 yes that is the exact functionality that I would like to support.  To provide some context on use case I would refer to the transaction prologue module [here](https://github.com/0xPolygonMiden/miden-base/blob/frisitano-prologue/miden-lib/asm/sat/prologue.masm#L1-L61).  We define a bunch of constants which are applicable across the transaction kernel context.  I would like to be able to define these in a shared constants module and then import them using the pattern you describe.  I don't think it would be too difficult to support this but maybe a matter of priorities.

**Comment by vlopes11** [comment]
It is not challenging, but might involve some work. I'll give some context here to document it for future reference, when we come to implement the change.

Currently, we assume the MASM source is self-resolvable and we create a local constants map from the file itself

https://github.com/0xPolygonMiden/miden-vm/blob/159a04f3d4648f77d664642409e521438378bd6c/assembly/src/parsers/mod.rs#L182-L190

However, at that point, the parser context holds only the imports path

https://github.com/0xPolygonMiden/miden-vm/blob/159a04f3d4648f77d664642409e521438378bd6c/assembly/src/parsers/mod.rs#L280-L302

It means we "open" the modules during compilation, via

https://github.com/0xPolygonMiden/miden-vm/blob/159a04f3d4648f77d664642409e521438378bd6c/assembly/src/assembler/instruction/procedures.rs#L63-L69

This code above will parse a module from the provider, and compile it into a set of exported procedures (code blocks)

Here is the module definition

https://github.com/0xPolygonMiden/miden-vm/blob/159a04f3d4648f77d664642409e521438378bd6c/assembly/src/library.rs#L234-L239

As you can see, holds only a link between a path (i.e. `std::math::u64_checked_add`) and a compited AST. The change to make constants works would have to make this link more complex, being a path to either a `ModuleAst` or a numeric constant.

Then, we would have to pass the context to the `push` compilers and query it for the constants:

https://github.com/0xPolygonMiden/miden-vm/blob/159a04f3d4648f77d664642409e521438378bd6c/assembly/src/assembler/instruction/env_ops.rs#L14-L34

**Comment by frisitano** [comment]
@vlopes11 thank you, that's really helpful context!  I think this is a little more involved then I had expected.  For now I think we're fine to repeat the definitions of the constants across modules but this isn't ideal.  When the team has capacity it would be great to introduce support for constant imports.

**Comment by bobbinth** [comment]
I think another challenge is that constants don't exist at the AST level. So, once a module is compiled, all information about constants is "erased". So, as the first thing, we'll probably want constants to be recorded in `ModuleAst` struct. For this, we'd want to figure out how to handle constant names (i.e., hash them as we do with procedure names or do something else?).

**Comment by bobbinth** [comment]
Superseded by #1290.
